### PR TITLE
Add cache deadlock test

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds a unit test for a deadlock issue that existed prior to
commit 1d0b93f76667e4af1bad7b0e522f1a4e6c8b3fbc. While commit
1d0b93f76667e4af1bad7b0e522f1a4e6c8b3fbc seems to have addressed the
deadlock issue it would be more robust to have a unit test for it and a
unit test will reduce the risk that future maintenance on Cache will
reintroduce the deadlock issue. This test reliably fails prior to but
passes after commit 1d0b93f76667e4af1bad7b0e522f1a4e6c8b3fbc.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/common/cache/CacheTests.java [hunk 1]
```diff
@@ -502,6 +505,91 @@ public void testComputeIfAbsentThrowsExceptionIfLoaderReturnsANullValue() {
...
+            Thread thread = new Thread(() -> {
+                Random random = new Random(random().nextLong());
+                latch.countDown();
+                for (int j = 0; j < numberOfEntries; j++) {
+                    Key key = new Key(random.nextInt(numberOfEntries));
+                    try {
+                        cache.computeIfAbsent(key, k -> k.key != 0 ? cache.get(new Key(k.key / 2)) : 0);
+                    } catch (ExecutionException e) {
+
+                    }
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/common/cache/CacheTests.java]: Is it fine to completely ignore it or could it hide bugs?

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/common/cache/CacheTests.java]: An exception is not expected here, so I agree that we should surface any exceptions that do appear here. I've added such an assertion in d840ea5e00218a05c1360da2814ff3d89816680e.

**Comment by jpountz** [comment]
LGTM

**Comment by jasontedor** [comment]
Thank you for reviewing @jpountz. I've squashed the two commits in this pull request and integrated into master.

**Pull request merged**